File: swift-bridge/Sources/ModelIOBridge/Asset.swift
import Foundation
import ModelIO
import simd

private func mdl_asset_info(_ asset: MDLAsset) -> [String: Any] {
    var info: [String: Any] = [
        "count": asset.count,
        "frame_interval": asset.frameInterval,
        "start_time": asset.startTime,
        "end_time": asset.endTime,
        "bounding_box": mdl_bounding_box(asset.boundingBox),
        "up_axis": [asset.upAxis.x, asset.upAxis.y, asset.upAxis.z],
    ]
    if let url = asset.url {
        info["url"] = url.path
    }
    return info
}

@_cdecl("mdl_asset_new_empty")
public func mdl_asset_new_empty(
    _ outAsset: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    mdl_run(outError) {
        guard let outAsset else {
            throw ModelIOBridgeError.invalidArgument("missing output asset pointer")
        }
        outAsset.pointee = mdl_retain(MDLAsset(bufferAllocator: nil))
    }
}

@_cdecl("mdl_asset_new_with_url")
public func mdl_asset_new_with_url(
    _ path: UnsafePointer<CChar>?,
    _ outAsset: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    mdl_run(outError) {
        guard let path, let outAsset else {
            throw ModelIOBridgeError.invalidArgument("missing asset path or output pointer")
        }
        let url = URL(fileURLWithPath: String(cString: path))
        outAsset.pointee = mdl_retain(MDLAsset(url: url))
    }
}

@_cdecl("mdl_asset_can_import_file_extension")
public func mdl_asset_can_import_file_extension(_ pathExtension: UnsafePointer<CChar>?) -> Int32 {
    guard let pathExtension else { return 0 }
    return MDLAsset.canImportFileExtension(String(cString: pathExtension)) ? 1 : 0
}

@_cdecl("mdl_asset_can_export_file_extension")
public func mdl_asset_can_export_file_extension(_ pathExtension: UnsafePointer<CChar>?) -> Int32 {
    guard let pathExtension else { return 0 }
    return MDLAsset.canExportFileExtension(String(cString: pathExtension)) ? 1 : 0
}

@_cdecl("mdl_asset_info_json")
public func mdl_asset_info_json(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset else { return nil }
    return mdl_string(mdl_json_string(from: mdl_asset_info(asset)) ?? "{}")
}

@_cdecl("mdl_asset_count")
public func mdl_asset_count(_ handle: UnsafeMutableRawPointer?) -> UInt64 {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset else { return 0 }
    return UInt64(asset.count)
}

@_cdecl("mdl_asset_bounding_box")
public func mdl_asset_bounding_box(
    _ handle: UnsafeMutableRawPointer?,
    _ outMinX: UnsafeMutablePointer<Float>?,
    _ outMinY: UnsafeMutablePointer<Float>?,
    _ outMinZ: UnsafeMutablePointer<Float>?,
    _ outMaxX: UnsafeMutablePointer<Float>?,
    _ outMaxY: UnsafeMutablePointer<Float>?,
    _ outMaxZ: UnsafeMutablePointer<Float>?
) {
    let zero = MDLAxisAlignedBoundingBox(maxBounds: SIMD3<Float>(repeating: 0), minBounds: SIMD3<Float>(repeating: 0))
    let boundingBox = (mdl_borrow_object(handle) as? MDLAsset)?.boundingBox ?? zero
    outMinX?.pointee = boundingBox.minBounds.x
    outMinY?.pointee = boundingBox.minBounds.y
    outMinZ?.pointee = boundingBox.minBounds.z
    outMaxX?.pointee = boundingBox.maxBounds.x
    outMaxY?.pointee = boundingBox.maxBounds.y
    outMaxZ?.pointee = boundingBox.maxBounds.z
}

@_cdecl("mdl_asset_bounding_box_at_time")
public func mdl_asset_bounding_box_at_time(
    _ handle: UnsafeMutableRawPointer?,
    _ time: Double,
    _ outMinX: UnsafeMutablePointer<Float>?,
    _ outMinY: UnsafeMutablePointer<Float>?,
    _ outMinZ: UnsafeMutablePointer<Float>?,
    _ outMaxX: UnsafeMutablePointer<Float>?,
    _ outMaxY: UnsafeMutablePointer<Float>?,
    _ outMaxZ: UnsafeMutablePointer<Float>?
) {
    let zero = MDLAxisAlignedBoundingBox(maxBounds: SIMD3<Float>(repeating: 0), minBounds: SIMD3<Float>(repeating: 0))
    let boundingBox = (mdl_borrow_object(handle) as? MDLAsset)?.boundingBox(atTime: time) ?? zero
    outMinX?.pointee = boundingBox.minBounds.x
    outMinY?.pointee = boundingBox.minBounds.y
    outMinZ?.pointee = boundingBox.minBounds.z
    outMaxX?.pointee = boundingBox.maxBounds.x
    outMaxY?.pointee = boundingBox.maxBounds.y
    outMaxZ?.pointee = boundingBox.maxBounds.z
}

@_cdecl("mdl_asset_url_string")
public func mdl_asset_url_string(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset,
          let url = asset.url
    else {
        return nil
    }
    return mdl_string(url.path)
}

@_cdecl("mdl_asset_object_at_index")
public func mdl_asset_object_at_index(
    _ handle: UnsafeMutableRawPointer?,
    _ index: UInt64
) -> UnsafeMutableRawPointer? {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset,
          index < UInt64(asset.count)
    else {
        return nil
    }
    return mdl_retain(asset.object(at: Int(index)))
}

@_cdecl("mdl_asset_mesh_at_index")
public func mdl_asset_mesh_at_index(
    _ handle: UnsafeMutableRawPointer?,
    _ index: UInt64
) -> UnsafeMutableRawPointer? {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset,
          index < UInt64(asset.count),
          let mesh = asset.object(at: Int(index)) as? MDLMesh
    else {
        return nil
    }
    return mdl_retain(mesh)
}

@_cdecl("mdl_asset_object_at_path")
public func mdl_asset_object_at_path(
    _ handle: UnsafeMutableRawPointer?,
    _ path: UnsafePointer<CChar>?
) -> UnsafeMutableRawPointer? {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset,
          let path
    else {
        return nil
    }
    return mdl_retain(asset.object(atPath: String(cString: path)))
}

@_cdecl("mdl_asset_add_object")
public func mdl_asset_add_object(_ handle: UnsafeMutableRawPointer?, _ objectHandle: UnsafeMutableRawPointer?) {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset,
          let object = mdl_borrow_object(objectHandle) as? MDLObject
    else {
        return
    }
    asset.add(object)
}

@_cdecl("mdl_asset_remove_object")
public func mdl_asset_remove_object(_ handle: UnsafeMutableRawPointer?, _ objectHandle: UnsafeMutableRawPointer?) {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset,
          let object = mdl_borrow_object(objectHandle) as? MDLObject
    else {
        return
    }
    asset.remove(object)
}

@_cdecl("mdl_asset_load_textures")
public func mdl_asset_load_textures(_ handle: UnsafeMutableRawPointer?) {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset else { return }
    asset.loadTextures()
}

@_cdecl("mdl_asset_export_to_url")
public func mdl_asset_export_to_url(
    _ handle: UnsafeMutableRawPointer?,
    _ path: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    mdl_run(outError) {
        guard let asset = mdl_borrow_object(handle) as? MDLAsset,
              let path
        else {
            throw ModelIOBridgeError.invalidArgument("missing asset or export path")
        }
        let url = URL(fileURLWithPath: String(cString: path))
        try asset.export(to: url)
    }
}

@_cdecl("mdl_asset_frame_interval")
public func mdl_asset_frame_interval(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset else { return 0 }
    return asset.frameInterval
}

@_cdecl("mdl_asset_set_frame_interval")
public func mdl_asset_set_frame_interval(_ handle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset else { return }
    asset.frameInterval = value
}

@_cdecl("mdl_asset_start_time")
public func mdl_asset_start_time(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset else { return 0 }
    return asset.startTime
}

@_cdecl("mdl_asset_set_start_time")
public func mdl_asset_set_start_time(_ handle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset else { return }
    asset.startTime = value
}

@_cdecl("mdl_asset_end_time")
public func mdl_asset_end_time(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset else { return 0 }
    return asset.endTime
}

@_cdecl("mdl_asset_set_end_time")
public func mdl_asset_set_end_time(_ handle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset else { return }
    asset.endTime = value
}

@_cdecl("mdl_asset_up_axis")
public func mdl_asset_up_axis(
    _ handle: UnsafeMutableRawPointer?,
    _ outX: UnsafeMutablePointer<Float>?,
    _ outY: UnsafeMutablePointer<Float>?,
    _ outZ: UnsafeMutablePointer<Float>?
) {
    let upAxis = (mdl_borrow_object(handle) as? MDLAsset)?.upAxis ?? SIMD3<Float>(0, 1, 0)
    outX?.pointee = upAxis.x
    outY?.pointee = upAxis.y
    outZ?.pointee = upAxis.z
}

@_cdecl("mdl_asset_set_up_axis")
public func mdl_asset_set_up_axis(_ handle: UnsafeMutableRawPointer?, _ x: Float, _ y: Float, _ z: Float) {
    guard let asset = mdl_borrow_object(handle) as? MDLAsset else { return }
    asset.upAxis = SIMD3<Float>(x, y, z)
}

File: src/asset.rs
use std::path::Path;
use std::ptr;

use crate::error::Result;
use crate::ffi;
use crate::handle::ObjectHandle;
use crate::mesh::Mesh;
use crate::object::Object;
use crate::types::{AssetInfo, BoundingBox};
use crate::util::{c_string, parse_json, path_to_c_string, required_handle, take_string};

#[derive(Debug, Clone)]
/// Wraps the corresponding Model I/O asset counterpart.
pub struct Asset {
    handle: ObjectHandle,
}

impl Asset {
    /// Builds this wrapper from the retained handle of the wrapped Model I/O asset counterpart.
    pub(crate) fn from_handle(handle: ObjectHandle) -> Self {
        Self { handle }
    }

    /// Returns the opaque pointer used to call the wrapped Model I/O asset counterpart.
    pub(crate) fn as_ptr(&self) -> *mut core::ffi::c_void {
        self.handle.as_ptr()
    }

    /// Wraps the corresponding Model I/O initializer for the wrapped Model I/O asset counterpart.
    pub fn new() -> Result<Self> {
        let mut out_asset = ptr::null_mut();
        let mut out_error = ptr::null_mut();
        // SAFETY: Output pointers are initialized and managed; FFI function is called safely.
        let status = unsafe { ffi::mdl_asset_new_empty(&mut out_asset, &mut out_error) };
        crate::util::status_result(status, out_error)?;
        Ok(Self::from_handle(required_handle(out_asset, "MDLAsset")?))
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn from_url(path: impl AsRef<Path>) -> Result<Self> {
        let path = path_to_c_string(path.as_ref())?;
        let mut out_asset = ptr::null_mut();
        let mut out_error = ptr::null_mut();
        let status =
            // SAFETY: Output pointers are initialized and managed; FFI function is called safely.
            unsafe { ffi::mdl_asset_new_with_url(path.as_ptr(), &mut out_asset, &mut out_error) };
        crate::util::status_result(status, out_error)?;
        Ok(Self::from_handle(required_handle(out_asset, "MDLAsset")?))
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn can_import_file_extension(path_extension: &str) -> bool {
        let trimmed = path_extension.trim_start_matches('.');
        crate::util::c_string(trimmed)
            // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
            .is_ok_and(|ext| unsafe { ffi::mdl_asset_can_import_file_extension(ext.as_ptr()) != 0 })
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn can_export_file_extension(path_extension: &str) -> bool {
        let trimmed = path_extension.trim_start_matches('.');
        crate::util::c_string(trimmed)
            // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
            .is_ok_and(|ext| unsafe { ffi::mdl_asset_can_export_file_extension(ext.as_ptr()) != 0 })
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn info(&self) -> Result<AssetInfo> {
        parse_json(
            // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
            unsafe { ffi::mdl_asset_info_json(self.handle.as_ptr()) },
            "MDLAsset",
        )
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn export_to_url(&self, path: impl AsRef<Path>) -> Result<()> {
        let path = path_to_c_string(path.as_ref())?;
        let mut out_error = ptr::null_mut();
        // SAFETY: The unsafe operation is valid in this context.
        let status = unsafe {
            ffi::mdl_asset_export_to_url(self.handle.as_ptr(), path.as_ptr(), &mut out_error)
        };
        crate::util::status_result(status, out_error)
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn count(&self) -> usize {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_asset_count(self.handle.as_ptr()) as usize }
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn bounding_box(&self) -> BoundingBox {
        let mut min = [0.0_f32; 3];
        let mut max = [0.0_f32; 3];
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_asset_bounding_box(
                self.handle.as_ptr(),
                &mut min[0],
                &mut min[1],
                &mut min[2],
                &mut max[0],
                &mut max[1],
                &mut max[2],
            );
        }
        BoundingBox { min, max }
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn bounding_box_at_time(&self, time: f64) -> BoundingBox {
        let mut min = [0.0_f32; 3];
        let mut max = [0.0_f32; 3];
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_asset_bounding_box_at_time(
                self.handle.as_ptr(),
                time,
                &mut min[0],
                &mut min[1],
                &mut min[2],
                &mut max[0],
                &mut max[1],
                &mut max[2],
            );
        }
        BoundingBox { min, max }
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn url(&self) -> Option<String> {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        take_string(unsafe { ffi::mdl_asset_url_string(self.handle.as_ptr()) })
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn frame_interval(&self) -> f64 {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_asset_frame_interval(self.handle.as_ptr()) }
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn set_frame_interval(&self, value: f64) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_asset_set_frame_interval(self.handle.as_ptr(), value) };
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn start_time(&self) -> f64 {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_asset_start_time(self.handle.as_ptr()) }
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn set_start_time(&self, value: f64) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_asset_set_start_time(self.handle.as_ptr(), value) };
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn end_time(&self) -> f64 {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_asset_end_time(self.handle.as_ptr()) }
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn set_end_time(&self, value: f64) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_asset_set_end_time(self.handle.as_ptr(), value) };
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn up_axis(&self) -> [f32; 3] {
        let mut axis = [0.0_f32; 3];
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_asset_up_axis(
                self.handle.as_ptr(),
                &mut axis[0],
                &mut axis[1],
                &mut axis[2],
            );
        };
        axis
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn set_up_axis(&self, axis: [f32; 3]) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_asset_set_up_axis(self.handle.as_ptr(), axis[0], axis[1], axis[2]) };
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn object_at(&self, index: usize) -> Option<Object> {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        let ptr = unsafe { ffi::mdl_asset_object_at_index(self.handle.as_ptr(), index as u64) };
        // SAFETY: The unsafe operation is valid in this context.
        unsafe { ObjectHandle::from_retained_ptr(ptr) }.map(Object::from_handle)
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn object_at_path(&self, path: &str) -> Result<Option<Object>> {
        let path = c_string(path)?;
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        let ptr = unsafe { ffi::mdl_asset_object_at_path(self.handle.as_ptr(), path.as_ptr()) };
        // SAFETY: The unsafe operation is valid in this context.
        Ok(unsafe { ObjectHandle::from_retained_ptr(ptr) }.map(Object::from_handle))
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn add_object(&self, object: &Object) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_asset_add_object(self.handle.as_ptr(), object.as_ptr()) };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn remove_object(&self, object: &Object) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_asset_remove_object(self.handle.as_ptr(), object.as_ptr()) };
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn mesh_at(&self, index: usize) -> Option<Mesh> {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        let ptr = unsafe { ffi::mdl_asset_mesh_at_index(self.handle.as_ptr(), index as u64) };
        // SAFETY: The unsafe operation is valid in this context.
        unsafe { ObjectHandle::from_retained_ptr(ptr) }.map(Mesh::from_handle)
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn meshes(&self) -> Vec<Mesh> {
        (0..self.count())
            .filter_map(|index| self.mesh_at(index))
            .collect()
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O asset counterpart.
    pub fn load_textures(&self) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_asset_load_textures(self.handle.as_ptr()) };
    }
}
